# Analysis on Stack Overflow 2023 Developer Survey

### The enclosed data set is the complete, cleaned results of the 2023 Stack Overflow Developer Survey. For more details on using this dataset, see [Stack Overflow](https://www.kaggle.com/datasets/stackoverflow/stack-overflow-2023-developers-survey?resource=download&select=survey_results_public.csv)




### This analysis answers five questions that draws insignts into some key aspects of professional developers. First, we import all necessary packages

In [1]:
install.packages("tidyverse")
library(tidyverse)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     


── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors


In [2]:
install.packages('dplyr')
library(dplyr)
install.packages("tidyr")
library(tidyr)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



## Loading the Dataset

In [3]:
survey_results <- read.csv("/kaggle/input/stackoverflow-survey/survey_results_public.csv")

In [4]:
head(survey_results)

ResponseId Q120    MainBranch                     Age            
1 1          I agree None of these                  18-24 years old
2 2          I agree I am a developer by profession 25-34 years old
3 3          I agree I am a developer by profession 45-54 years old
4 4          I agree I am a developer by profession 25-34 years old
5 5          I agree I am a developer by profession 25-34 years old
6 6          I agree I am a developer by profession 35-44 years old
  Employment                                                              
1 NA                                                                      
2 Employed, full-time                                                     
3 Employed, full-time                                                     
4 Employed, full-time                                                     
5 Employed, full-time;Independent contractor, freelancer, or self-employed
6 Employed, full-time                                                     
  RemoteWork                          
1 NA                                  
2 Remote                              
3 Hybrid (some remote, some in-person)
4 Hybrid (some remote, some in-person)
5 Remote                              
6 Remote                              
  CodingActivities                                                                                                                     
1 NA                                                                                                                                   
2 Hobby;Contribute to open-source projects;Bootstrapping a business;Professional development or self-paced learning from online courses
3 Hobby;Professional development or self-paced learning from online courses                                                            
4 Hobby                                                                                                                                
5 Hobby;Contribute to open-source projects;Professional development or self-paced learning from online courses                         
6 Hobby;Professional development or self-paced learning from online courses                                                            
  EdLevel                                               
1 NA                                                    
2 Bachelor’s degree (B.A., B.S., B.Eng., etc.)          
3 Bachelor’s degree (B.A., B.S., B.Eng., etc.)          
4 Bachelor’s degree (B.A., B.S., B.Eng., etc.)          
5 Bachelor’s degree (B.A., B.S., B.Eng., etc.)          
6 Some college/university study without earning a degree
  LearnCode                                                                                                                                                                                                                                 
1 NA                                                                                                                                                                                                                                        
2 Books / Physical media;Colleague;Friend or family member;Hackathons (virtual or in-person);Online Courses or Certification;On the job training;Other online resources (e.g., videos, blogs, forum);School (i.e., University, College, etc)
3 Books / Physical media;Colleague;On the job training;Other online resources (e.g., videos, blogs, forum);School (i.e., University, College, etc)                                                                                          
4 Colleague;Friend or family member;Other online resources (e.g., videos, blogs, forum);School (i.e., University, College, etc)                                                                                                             
5 Books / Physical media;Online Courses or Certification;Other online resources (e.g., videos, blogs, forum);School (i.e., University, College, etc)                                            

## Creating Dataframes

### To carry out the analysis, we would create five dataframes that contain relevant columns for all five questions in our analysis

In [5]:
first_question <- select(survey_results, MainBranch, Employment, DevType, CompTotal, Industry)
second_question <- select(survey_results, RemoteWork, Employment, Age, OrgSize, Country, CodingActivities)
third_question <- select(survey_results, CompTotal, YearsCode, YearsCodePro)
forth_question <- select(survey_results, LearnCode)
fifth_question <- select(survey_results, Employment, MainBranch, EdLevel)

## First Question

### What industries do developers get paid the most?

In [6]:
glimpse(first_question)

Rows: 89,184
Columns: 5
$ MainBranch <chr> "None of these", "I am a developer by profession", "I am a …
$ Employment <chr> NA, "Employed, full-time", "Employed, full-time", "Employed…
$ DevType    <chr> NA, "Senior Executive (C-Suite, VP, etc.)", "Developer, bac…
$ CompTotal  <dbl> NA, 285000, 250000, 156000, 1320000, 78000, 135000, 80000, …
$ Industry   <chr> NA, "Information Services, IT, Software Development, or oth…


In [7]:
# First, specify data for only professionals as developers

developer_professionals <- first_question %>% filter(MainBranch == "I am a developer by profession")

In [8]:
#From the data, CompTotal covers several currencies. To get an accurate compensation average for each industry, we have to make the currency consistent

#For this question, we can convert only currencies used atleast 100 times. Let's see the most used currencies from highest to lowest

currencyy_count <- survey_results %>% group_by(Currency) %>% summarise(count = n()) %>% arrange(desc(count))
currencyy_count <- na.omit(currencyy_count)
View(currencyy_count)

Currency                    count
1   EUR European Euro           17651
2   USD\tUnited States dollar   16729
3   GBP\tPound sterling          4473
4   INR\tIndian rupee            3615
5   CAD\tCanadian dollar         2647
6   PLN\tPolish zloty            1606
7   AUD\tAustralian dollar       1594
8   BRL\tBrazilian real          1475
9   SEK\tSwedish krona           1324
10  CHF\tSwiss franc              889
11  ILS\tIsraeli new shekel       695
12  DKK\tDanish krone             677
13  RUB\tRussian ruble            674
14  CZK\tCzech koruna             672
15  NOK\tNorwegian krone          580
16  NZD\tNew Zealand dollar       531
17  TRY\tTurkish lira             525
18  RON\tRomanian leu             468
19  ZAR\tSouth African rand       436
20  UAH\tUkrainian hryvnia        408
21  HUF\tHungarian forint         393
22  CNY\tChinese Yuan Renminbi    388
23  MXN\tMexican peso             383
24  PKR\tPakistani rupee          360
25  IRR\tIranian rial             344
26  BGN\tBulgarian lev            339
27  BDT\tBangladeshi taka         279
28  JPY\tJapanese yen             279
29  COP\tColombian peso           273
30  IDR\tIndonesian rupiah        272
⋮   ⋮                           ⋮    
115 NAD\tNamibian dollar        3    
116 SLL\tSierra Leonean leone   3    
117 SRD\tSurinamese dollar      3    
118 SZL\tSwazi lilangeni        3    
119 TJS\tTajikistani somoni     3    
120 XPF\tCFP franc              3    
121 BBD\tBarbadian dollar       2    
122 BMD\tBermudian dollar       2    
123 BWP\tBotswana pula          2    
124 CDF\tCongolese franc        2    
125 ERN\tEritrean nakfa         2    
126 FKP\tFalkland Islands pound 2    
127 GIP\tGibraltar pound        2    
128 GYD\tGuyanese dollar        2    
129 HRK\tCroatian kuna          2    
130 LSL\tLesotho loti           2    
131 SOS\tSomali shilling        2    
132 TMT\tTurkmen manat          2    
133 TOP\tTongan pa’anga         2    
134 XCD\tEast Caribbean dollar  2    
135 none\tFaroese krona         2    
136 CVE\tCape Verdean escudo    1    
137 DJF\tDjiboutian franc       1    
138 JEP\tJersey pound           1    
139 KMF\tComorian franc         1    
140 KYD\tCayman Islands dollar  1    
141 LRD\tLiberian dollar        1    
142 MOP\tMacanese pataca        1    
143 SDG\tSudanese pound         1    
144 SSP\tSouth Sudanese pound   1

In [9]:
#Create a list of USD exchange_rates for the top currencies

exchange_rates <- c(
  "EUR" = 1.1,  # Euro to USD
  "GBP" = 1.3,  # British Pound to USD
  "INR" = 0.012,  # Indian Rupee to USD
  "USD" = 1,  # USD stays the same
  "JPY" = 0.007, # Japanese Yen to USD
  "CAD" = 0.75,
  "PLN" = 0.26,
  "AUD" = 0.67,
  "BRL" = 0.18,
  "SEK" = 0.097,
  "CHF" = 1.19,
  "ILS" = 0.27,
  "DKK" = 0.15,
  "RUB" = 0.011,
  "CZK" = 0.044,
  "NOK" = 0.093,
  "NZD" = 0.62,
  "TRY" = 0.03,
  "RON" = 0.22,
  "ZAR" = 0.056,
  "UAH" = 0.024,
  "HUF" = 0.0028,
  "CNY" = 0.14,
  "MXN" = 0.05,
  "PKR" = 0.0036,
  "IRR" = 0.000024,
  "BGN" = 0.57,
  "BDT" = 0.0084,
  "COP" = 0.00024,
  "IDR" = 0.000065,
  "ARS" = 0.0010,
  "PHP" = 0.018,
  "SGD" = 0.77,
  "RSD" = 0.0095,
  "VND" = 0.000041,
  "KRW" = 0.00075,
  "MYR" = 0.230548,
  "AED" = 0.27,
  "TWD" = 0.031,
  "CLP" = 0.0011,
  "NGN" = 0.00062,
  "HKD" = 0.13,
  "LKR" = 0.0033,
  "THB" = 0.030,
  "EGP" = 0.021,
  "KES" = 0.0078,
  "NPR" = 0.0074
)

In [10]:
#Now, convert the currencies listed above to USD and save in a new column "compensation_USD" 
developer_professionals <- survey_results %>%
    mutate(currency_short = substr(Currency, 1, 3)) %>% 
    mutate(compensation_USD = CompTotal * exchange_rates[currency_short])

#Then, we calculate industry that pays the most.
industry_compensation <- developer_professionals %>% 
  group_by(Industry) %>% 
  summarise(avg_compensation = mean(compensation_USD, na.rm = TRUE)) %>%
  arrange(desc(avg_compensation))


View(industry_compensation)

Industry                                                           
1  NA                                                                 
2  Information Services, IT, Software Development, or other Technology
3  Financial Services                                                 
4  Other                                                              
5  Advertising Services                                               
6  Healthcare                                                         
7  Insurance                                                          
8  Retail and Consumer Services                                       
9  Manufacturing, Transportation, or Supply Chain                     
10 Legal Services                                                     
11 Oil & Gas                                                          
12 Wholesale                                                          
13 Higher Education                                                   
   avg_compensation
1  6.739554e+41    
2  7.548880e+11    
3  1.277960e+05    
4  1.265913e+05    
5  1.122970e+05    
6  1.057785e+05    
7  1.017589e+05    
8  9.987641e+04    
9  9.123980e+04    
10 9.048492e+04    
11 8.994258e+04    
12 7.407975e+04    
13 7.378497e+04


* ### "Information Services, IT, Software Developent, or other technology" is the highest paying industry.
* ### For a more accurate result, we can convert every currency to USD.
* ### There is a large value for "NA" because a lot of industries are null in the survey (52410 out of 89184). See the code below.

In [11]:
na_count <- sum(is.na(survey_results$Industry))
head(na_count)

[1] 52410

## Second Question

### How much does remote working matter to employees?

In [12]:
glimpse(second_question)

Rows: 89,184
Columns: 6
$ RemoteWork       <chr> NA, "Remote", "Hybrid (some remote, some in-person)",…
$ Employment       <chr> NA, "Employed, full-time", "Employed, full-time", "Em…
$ Age              <chr> "18-24 years old", "25-34 years old", "45-54 years ol…
$ OrgSize          <chr> NA, "2 to 9 employees", "5,000 to 9,999 employees", "…
$ Country          <chr> NA, "United States of America", "United States of Ame…
$ CodingActivities <chr> NA, "Hobby;Contribute to open-source projects;Bootstr…


In [13]:
# First, filter the dataframe to only include employees before carrying out analysis

remote_work <- second_question %>%
    filter(Employment %in% 
    c("Employed, full-time", "Employed, part-time", "Independent contractor, freelancer, or self-employed"))

remote_happiness <- remote_work %>%
  group_by(RemoteWork) %>%
  summarize(count = n()) %>%
  arrange(desc(count)) %>%
  mutate(percentage = round(count / sum(count) * 100, 2))

View(remote_happiness)

RemoteWork                           count percentage
1 Hybrid (some remote, some in-person) 26537 42.48     
2 Remote                               25614 41.00     
3 In-person                            10304 16.49     
4 NA                                      13  0.02


* ### From the result, about 83% of all employees work remotely in some form, with 41% fully remote workers. As such, remote work matters a lot.

## Third Question

### How does coding experience affect level of pay

In [14]:
glimpse(third_question)

Rows: 89,184
Columns: 3
$ CompTotal    <dbl> NA, 285000, 250000, 156000, 1320000, 78000, 135000, 80000…
$ YearsCode    <chr> NA, "18", "27", "12", "6", "21", "4", "5", "20", "6", "14…
$ YearsCodePro <chr> NA, "9", "23", "7", "4", "21", "3", "3", "15", NA, "3", "…


In [15]:
earning <- third_question %>%
  group_by(YearsCode) %>%
  summarize(
    count = n(),
    mean_compensation = mean(CompTotal, na.rm = TRUE)
  ) %>%
  arrange(desc(mean_compensation))


View(earning)

YearsCode          count mean_compensation
1  More than 50 years  353  4.504505e+44     
2  11                 2519  5.865103e+17     
3  22                 1210  1.216545e+13     
4  8                  4879  3.612733e+12     
5  10                 6521  2.504513e+11     
6  4                  4466  3.343340e+09     
7  Less than 1 year    968  3.372115e+08     
8  NA                 1749  2.678489e+08     
9  36                  282  6.353451e+07     
10 3                  4269  3.108499e+07     
11 17                 1544  2.761786e+07     
12 14                 2278  2.484318e+07     
13 32                  444  2.267346e+07     
14 12                 3471  2.159428e+07     
15 9                  3130  2.147536e+07     
16 5                  5415  2.131504e+07     
17 7                  4800  1.922351e+07     
18 18                 1690  1.486990e+07     
19 19                  826  1.112820e+07     
20 15                 4336  1.048651e+07     
21 2                  2826  9.629769e+06     
22 6                  4893  9.296070e+06     
23 24                  959  8.192939e+06     
24 16                 1842  5.987600e+06     
25 21                  761  5.926232e+06     
26 1                  1294  5.503270e+06     
27 13                 2601  4.204583e+06     
28 20                 4021  2.885957e+06     
29 25                 2685  2.199604e+06     
30 26                  706  1.655409e+06     
31 23                 1327  1.642028e+06     
32 27                  710  1.402850e+06     
33 35                 1118  1.400847e+06     
34 30                 2078  1.353212e+06     
35 28                  715  1.174558e+06     
36 31                  272  1.088323e+06     
37 45                  360  8.555072e+05     
38 38                  428  8.460540e+05     
39 33                  496  5.800757e+05     
40 39                  229  5.057700e+05     
41 34                  333  4.535715e+05     
42 29                  324  4.236405e+05     
43 44                  160  4.232375e+05     
44 37                  312  4.051481e+05     
45 40                 1302  3.990705e+05     
46 43                  265  2.915439e+05     
47 42                  347  2.801256e+05     
48 48                   96  2.513360e+05     
49 41                  227  2.387978e+05     
50 49                   42  2.045238e+05     
51 47                   79  1.515690e+05     
52 46                  101  1.363242e+05     
53 50                  125  1.346256e+05

* ### People with over 50 years professional coding experience earn the highest. This might suggest employees within this bracket are in managerial positions or owners of companies.
* ### But aside that, there's no particular pattern that would suggest experience affects pay.

## Forth Question

### What's the most popular method of learning to code

In [16]:
glimpse(forth_question)

Rows: 89,184
Columns: 1
$ LearnCode <chr> NA, "Books / Physical media;Colleague;Friend or family membe…


In [17]:
#From the dataset, respondents were given the opportunity to select as many learning methods as possible from the 10 specified options. 
#As such, the the LearnCode column contains more than one entry, seperated with a semi-colon.
# To get the most popular learning method, We seperate each learning option into a different column

seperate_data <- forth_question %>% 
    separate(LearnCode, into = c("resource_1", "resource_2", "resource_3", "resource_4", "resource_5", "resource_6", "resource_7", "resource_8","resource_9", "resource_10"), sep = ";", fill = "right")


head(seperate_data)

resource_1             resource_2                     
1 NA                     NA                             
2 Books / Physical media Colleague                      
3 Books / Physical media Colleague                      
4 Colleague              Friend or family member        
5 Books / Physical media Online Courses or Certification
6 Books / Physical media Colleague                      
  resource_3                                         
1 NA                                                 
2 Friend or family member                            
3 On the job training                                
4 Other online resources (e.g., videos, blogs, forum)
5 Other online resources (e.g., videos, blogs, forum)
6 Online Courses or Certification                    
  resource_4                                         
1 NA                                                 
2 Hackathons (virtual or in-person)                  
3 Other online resources (e.g., videos, blogs, forum)
4 School (i.e., University, College, etc)            
5 School (i.e., University, College, etc)            
6 Other online resources (e.g., videos, blogs, forum)
  resource_5                              resource_6         
1 NA                                      NA                 
2 Online Courses or Certification         On the job training
3 School (i.e., University, College, etc) NA                 
4 NA                                      NA                 
5 NA                                      NA                 
6 NA                                      NA                 
  resource_7                                         
1 NA                                                 
2 Other online resources (e.g., videos, blogs, forum)
3 NA                                                 
4 NA                                                 
5 NA                                                 
6 NA                                                 
  resource_8                              resource_9 resource_10
1 NA                                      NA         NA         
2 School (i.e., University, College, etc) NA         NA         
3 NA                                      NA         NA         
4 NA                                      NA         NA         
5 NA                                      NA         NA         
6 NA                                      NA         NA

In [18]:
# Convert the dataframe "seperated_data" from wide data to long data

seperate_data <- seperate_data %>% 
    pivot_longer(names_to = "Resource", values_to = "correct", resource_1:resource_10)

head(seperate_data)

Resource   correct
1 resource_1 NA     
2 resource_2 NA     
3 resource_3 NA     
4 resource_4 NA     
5 resource_5 NA     
6 resource_6 NA

In [19]:
#Determine the count for all learning method

top_learning_avenue <- seperate_data %>% 
  group_by(correct) %>% 
  summarize(count = n()) %>% 
  arrange(desc(count))

View(top_learning_avenue)

correct                                             count 
1  NA                                                  597107
2  Other online resources (e.g., videos, blogs, forum)  70244
3  Books / Physical media                               45406
4  School (i.e., University, College, etc)              43957
5  Online Courses or Certification                      43201
6  On the job training                                  40380
7  Colleague                                            20523
8  Friend or family member                               9936
9  Coding Bootcamp                                       8602
10 Hackathons (virtual or in-person)                     7033
11 Other (please specify):                               5451

## Fifth Question

### Are you more likely to get a job as a developer if you have a master's degree?

In [20]:
glimpse(fifth_question)

Rows: 89,184
Columns: 3
$ Employment <chr> NA, "Employed, full-time", "Employed, full-time", "Employed…
$ MainBranch <chr> "None of these", "I am a developer by profession", "I am a …
$ EdLevel    <chr> NA, "Bachelor’s degree (B.A., B.S., B.Eng., etc.)", "Bachel…


In [21]:
#For this, we evaluate the total number of masters degree holders to the number of employed masters degree holders.
#First, filter data for only employed developers respondents and get number of employed for each educational level

masters_preference <- fifth_question %>% 
  filter(Employment %in% c("Employed, full-time", "Employed, part-time", "Independent contractor, freelancer, or self-employed"), MainBranch == "I am a developer by profession")

masters_preference <- masters_preference %>%
  group_by(EdLevel) %>%
  summarize(count_employed = n())

View(masters_preference)

EdLevel                                                                           
1 Associate degree (A.A., A.S., etc.)                                               
2 Bachelor’s degree (B.A., B.S., B.Eng., etc.)                                      
3 Master’s degree (M.A., M.S., M.Eng., MBA, etc.)                                   
4 Primary/elementary school                                                         
5 Professional degree (JD, MD, Ph.D, Ed.D, etc.)                                    
6 Secondary school (e.g. American high school, German Realschule or Gymnasium, etc.)
7 Some college/university study without earning a degree                            
8 Something else                                                                    
  count_employed
1  1710         
2 26145         
3 14904         
4   307         
5  2253         
6  2468         
7  6165         
8   639

In [22]:
# Now, find the total number of respondents for each educational level, regardless of employment status

general <- fifth_question %>%
  group_by(EdLevel) %>%
  summarize(count_general = n())

View(general)

EdLevel                                                                           
1 Associate degree (A.A., A.S., etc.)                                               
2 Bachelor’s degree (B.A., B.S., B.Eng., etc.)                                      
3 Master’s degree (M.A., M.S., M.Eng., MBA, etc.)                                   
4 Primary/elementary school                                                         
5 Professional degree (JD, MD, Ph.D, Ed.D, etc.)                                    
6 Secondary school (e.g. American high school, German Realschule or Gymnasium, etc.)
7 Some college/university study without earning a degree                            
8 Something else                                                                    
9 NA                                                                                
  count_general
1  2807        
2 36706        
3 20543        
4  1905        
5  3887        
6  8897        
7 11753        
8  1475        
9  1211

In [23]:
#Join both tables together

likelihood <- merge(masters_preference, general, by='EdLevel')
View(likelihood)

EdLevel                                                                           
1 Associate degree (A.A., A.S., etc.)                                               
2 Bachelor’s degree (B.A., B.S., B.Eng., etc.)                                      
3 Master’s degree (M.A., M.S., M.Eng., MBA, etc.)                                   
4 Primary/elementary school                                                         
5 Professional degree (JD, MD, Ph.D, Ed.D, etc.)                                    
6 Secondary school (e.g. American high school, German Realschule or Gymnasium, etc.)
7 Some college/university study without earning a degree                            
8 Something else                                                                    
  count_employed count_general
1  1710           2807        
2 26145          36706        
3 14904          20543        
4   307           1905        
5  2253           3887        
6  2468           8897        
7  6165          11753        
8   639           1475

In [24]:
#Calculate the percentage likelihood of getting employed for each educational level.

likelihood <- likelihood %>%
    mutate(percentage_likelihood = round(count_employed / count_general * 100, 2)) %>%
    arrange(desc(percentage_likelihood))

View(likelihood)

EdLevel                                                                           
1 Master’s degree (M.A., M.S., M.Eng., MBA, etc.)                                   
2 Bachelor’s degree (B.A., B.S., B.Eng., etc.)                                      
3 Associate degree (A.A., A.S., etc.)                                               
4 Professional degree (JD, MD, Ph.D, Ed.D, etc.)                                    
5 Some college/university study without earning a degree                            
6 Something else                                                                    
7 Secondary school (e.g. American high school, German Realschule or Gymnasium, etc.)
8 Primary/elementary school                                                         
  count_employed count_general percentage_likelihood
1 14904          20543         72.55                
2 26145          36706         71.23                
3  1710           2807         60.92                
4  2253           3887         57.96                
5  6165          11753         52.45                
6   639           1475         43.32                
7  2468           8897         27.74                
8   307           1905         16.12

### Masters degree holders are more likely to become employed with 72.55%
### We can gain tons of insights from this dataset. Let me know if you'll like to add a question 😊